framework/src/node/state_machine/state_machine.ts

```diff
@@ -0,0 +1,241 @@
+/*
+ * Copyright © 2021 Lisk Foundation
+ *
+ * See the LICENSE file at the top-level directory of this distribution
+ * for licensing information.
+ *
+ * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
+ * no part of this software, including this file, may be copied, modified,
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ */
+
+import { Schema } from '@liskhq/lisk-codec';
+import { BlockContext } from './block_context';
+import { GenesisBlockContext } from './genesis_block_context';
+import { TransactionContext } from './transaction_context';
+import {
+	BlockExecuteContext,
+	GenesisBlockExecuteContext,
+	TransactionVerifyContext,
+	TransactionExecuteContext,
+	VerifyStatus,
+	BlockVerifyContext,
+	VerifycationResult,
+	CommandVerifyContext,
+	CommandExecuteContext,
+} from './types';
+
+export interface StateMachineCommand {
+	id: number;
+	schema: Schema;
+	verify?: <T = unknown>(ctx: CommandVerifyContext<T>) => Promise<VerifycationResult>;
+	execute: <T = unknown>(ctx: CommandExecuteContext<T>) => Promise<void>;
+}
+
+export interface StateMachineModule {
+	id: number;
+	commands: StateMachineCommand[];
+	verifyTransaction?: (ctx: TransactionVerifyContext) => Promise<VerifycationResult>;
+	afterGenesisBlockExecute?: (ctx: GenesisBlockExecuteContext) => Promise<void>;
+	verifyBlock?: (ctx: BlockVerifyContext) => Promise<void>;
+	beforeBlockExecute?: (ctx: BlockExecuteContext) => Promise<void>;
+	afterBlockExecute?: (ctx: BlockExecuteContext) => Promise<void>;
+	beforeTransactionExecute?: (ctx: TransactionExecuteContext) => Promise<void>;
+	afterTransactionExecute?: (ctx: TransactionExecuteContext) => Promise<void>;
+}
+
+export class StateMachine {
+	private readonly _modules: StateMachineModule[] = [];
+	private readonly _systemModules: StateMachineModule[] = [];
+
+	public registerModule(mod: StateMachineModule): void {
```

Comment:
yea i was thinking when application/node registerModule is called, they can call this method. application/node registerModule will need to call generator.registerModule as well

Reply:
from another angle, since state machine is responsible for transitioning the state, its better that it manages everything? from registering to execution?

Reply:
regarding the usage of variable `mod`, not convincing though, we could use `module` more expressive and its short already

Reply:
`module` is reserved word in node, so it's better not to use

Reply:
Also, the module will need to be registered to the generator and broker

Reply:
we do the validation of a module being registered in `node.ts`, and I was saying if we can do,
`new StateMachine({modules:  this._registeredModules })` in `node.ts` as I see unnecessary calling of registerModule on both generator and stateMachine, we are only doing the validation for duplicate registration of a module here.

Reply:
I was thinking the `StateMachine` will be instantiated before all module are registered. 
for example, when application is created, `Application`, `Node`, `StateMachine`, `ValidatorModule` and `BFTModule` etc will be instantiated so that other module can access validatorAPI or bftAPI for adding dependency.
If we instantiate everything at the same time, using constructor will be difficult

Reply:
ah okay I get your point, it will create instantiation sequence complexity. yeah then maybe the current approach is fine